proposals/ref-out-lambda-params.md

```diff
@@ -0,0 +1,82 @@
+# Declaration of lambda parameters with by-reference modifiers without type name
+
+## Summary  
+
+Allow lambda parameter declarations with by-reference modifiers (`in` / `ref` / `out` / `ref readonly`) to be declared without requiring their type names.
+
+Given this delegate:
+```cs
+delegate bool TryParse<T>(string text, out T result);
+```
+
+Allow this simplified parameter declaration:
+```cs
+TryParse<int> parse1 = (text, out result) => Int32.TryParse(text, out result);
+```
+
+Currently only this is valid:
+```cs
+TryParse<int> parse2 = (string text, out int result) => Int32.TryParse(text, out result);
+```
+
+## Detailed design
+
+### Parameter declaration
+
+Parameter declarations in lambda expressions with parenthesized parameters now permit a single identifier after a modifier on the parameter. This does not apply to lambda expressions with a single parameter with omitted parentheses.
+
+For example,
+```csharp
+SelfReturnerIn<string> fin = in x => x;
+SelfReturnerRef<string> fref = ref x => x;
+SelfReturnerOut<string> fout = out x => x;
+SelfReturnerRefReadonly<string> frr = ref readonly x => x;
+
+delegate T SelfReturnerIn<T>(in T t);
+delegate T SelfReturnerRef<T>(ref T t);
+delegate T SelfReturnerOut<T>(out T t);
+delegate T SelfReturnerRefReadonly<T>(ref readonly T t);
+```
+
+are all illegal, due to ambiguity with taking the reference of the returned expression in the `ref` case. For consistency, `in`, `out` and `ref readonly` are also left unsupported and illegal.
+
+The change in the spec will require that, [the grammar for lambda expressions](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/expressions#12191-general) must be adjusted as follows:
+
+```diff
+  implicit_anonymous_function_parameter_list
+-     : implicit_anonymous_function_parameter
+-       (',' implicit_anonymous_function_parameter)*
++     : implicit_parenthesized_anonymous_function_parameter
++       (',' implicit_parenthesized_anonymous_function_parameter)*
+      ;
+
++ implicit_parenthesized_anonymous_function_parameter
++     : anonymous_function_parameter_modifier? identifier
+      ;
+```
+
+The type of the parameters matches the type of the parameter in the target delegate type, including the by-reference modifiers.
+
+Attributes on the parameters will not be affected in any way. Similarly, `async` lambdas will also not be affected from this change.
```

Comment:
I'm not sure what "Similarly, `async` lambdas will also not be affected from this change." means here. They will be affected by this change; it still will be illegal to have `RefFunc<int, Task> f = async (ref i) => { ... };`, but not because the grammar forbids the syntax, it's a semantic error just like having async method take ref parameters is.